Simplified widget behavior

AnOctopus opened the pull request:
Recreating this PR, because I think I will use the prototype as a starting point for the final implementation of making widgets behave according to the model we decided on, so I want this to be available if I ask people for opinions/review

AnOctopus committed:
Simple widget-state model prototype

Keyed widgets act like unkeyed ones, resetting when args change. Session
state api for get/set their values still works, but was surprisingly
high effort to implement.

AnOctopus committed:
Handle old widget values for reset keyed widgets

AnOctopus committed:
Unused import cleanup

AnOctopus committed:
Experiment with explicit dict stack management

Removing the eager conversion of user keys into matching widget ids when
they are known, and so avoiding some bookkeeping.

By making `__getitem__` do the appropriate lookup for any combination of
user key and widget id, lots of callsites are simplified, and combined
with implementing `keys` to do lazy normalization to widget ids, much of
the code that previously needed to know about dict ordering and keys vs
ids can instead be written without it.

AnOctopus committed:
Mark slow unit tests with pytest

This makes them skippable with `-m='not slow'`

AnOctopus committed:
SessionState and widget notes and cleanup

Separated out from hypothesis test changes

AnOctopus committed:
Property tests and SessionState test methods

Very in progress, especially the strategy for session state which is
pretty hacky. I hope to have this in a good state very soon, so that it
can give me much more confidence in refactoring.

AnOctopus committed:
Rename `_compact_state` to `_compact`

It acting on state is implied, as it is a method on SessionState.

AnOctopus committed:
Remove incorrect conversion to widget id in state setting

AnOctopus committed:
Remove unused imports

AnOctopus committed:
Rename `.set_from_proto` -> `.set_widgets_from_proto`

AnOctopus committed:
Replace session_state strategy

New one is a terrible mess that needs refactoring and generates some
invalid session state instances, but it generates instances that
have keyed widgets and session state interacting with that, so it
exercises the code much better.

AnOctopus committed:
Adding better session state internal api methods

Working towards a construction/setting api that enforces more invariants
and is higher level than the really basic existing methods. Motivation
for this is the difficulty revealed by trying to generate useful
instances for hypothesis property tests.

AnOctopus committed:
Move widget metadata setting into widget methods

AnOctopus committed:
Clean up session state strategy

AnOctopus committed:
Minor strategy cleanup

AnOctopus committed:
Fix potential exception in `session_state.items()`

AnOctopus committed:
Hide function attributes in WidgetMetadata repr

They convey no real information, but add a lot of noise when looking at
it in logs.

AnOctopus committed:
Add compacting and second setting round to state gen

AnOctopus committed:
Remove unused strategies

AnOctopus committed:
Remove intermediate list in WState methods

AnOctopus committed:
Normalize wid and key parameter order

AnOctopus committed:
Simplify getting value for registration

AnOctopus committed:
reverse_key_wid_map property for simplification

AnOctopus committed:
Merge branch 'develop' into widgets-simplified-prototype

AnOctopus committed:
Help mypy know what `self` is in SessionState method

AnOctopus committed:
Add hypothesis as dev package

AnOctopus committed:
Fix session state e2e tests

Keyed widget behavior has deliberately changed, so tests updated to
reflect that.

Filtering out internal streamlit keys in session state.

vdonato commented on lib/streamlit/state/widgets.py:
```diff
@@ -268,16 +271,13 @@ def _get_widget_id(
     """Generate a widget id for the given widget.
 
-    If user_key is defined, the widget_id returned is simply user_key.
+    If user_key is defined, the widget_id includes it so keys can make widgets with identical arguments distinct.
```
super nit: line should wrap at ~80 characters (same comment for a few other long comments/docstrings)

vdonato commented on lib/tests/streamlit/delta_generator_test.py:
```diff
@@ -181,17 +181,6 @@ def test_duplicate_widget_id_error(self):
-                ),
-                str(ctx.exception),
-            )
-
```
With these lines gone, the `st.button` created in L168 isn't necessary anymore.

I do think it'd probably still be useful to test the keyed case here too by making this code match the code above aside from the presence of a non-`None` key parameter (perhaps this would be most cleanly done via `pytest.mark.parametrize`)

vdonato commented on lib/tests/streamlit/delta_generator_test.py:
```diff
@@ -531,26 +520,6 @@ def test_ids_are_diff_when_types_are_diff(self):
 
 
 class KeyWidgetIdTests(testutil.DeltaGeneratorTestCase):
-    def test_ids_are_equal_when_keys_are_equal(self):
-        text_input1 = TextInput()
-        text_input1.label = "Label #1"
-        text_input1.default = "Value #1"
-
-        text_input2 = TextInput()
-        text_input2.label = "Label #2"
-        text_input2.default = "Value #2"
-
-        element1 = Element()
-        element1.text_input.CopyFrom(text_input1)
-
-        element2 = Element()
-        element2.text_input.CopyFrom(text_input2)
-
-        register_widget("text_input", element1.text_input, user_key="some_key")
-
-        with self.assertRaises(DuplicateWidgetID):
-            register_widget("text_input", element2.text_input, user_key="some_key")
```
It'd also be useful to verify that no `DuplicateWidgetID` exception gets thrown when

* all non-key parameters are identical but the key parameters differ
* the key parameters are the same but there's at least one differing non-key parameter

The error cases can all be covered by `test_duplicate_widget_id_error` above

vdonato commented on lib/tests/streamlit/legacy_caching/hashing_test.py:
```diff
@@ -384,6 +385,7 @@ def test_file_position(self):
             f.seek(0)
             self.assertEqual(h1, get_hash(f))
 
+    @pytest.mark.skip
```
Was this added due to some dependency breakage that's now fixed in `develop`?

vdonato commented on lib/streamlit/state/session_state.py:
```diff
@@ -147,14 +140,12 @@ def keys(self) -> Set[str]:
+        return {self[wid] for wid in self}
 
-    def set_from_proto(self, widget_state: WidgetStateProto):
+    def set_widgets_from_proto(self, widget_state: WidgetStateProto):
```
nit: should this function be `set_widget_from_proto` (with the function in the SessionState calling this one keeping the plural name)

vdonato commented on lib/streamlit/state/session_state.py:
```diff
@@ -260,23 +251,36 @@ class SessionState(MutableMapping[str, Any]):
 
+    def _compact(self) -> "SessionState":
+        state: SessionState = self.copy()
+        state.compact_state
```
Looks like the parens to call `compact_state` are missing here

vdonato commented on lib/tests/streamlit/state/session_state_test.py:
```diff
@@ -692,3 +646,58 @@ def test_setattr_reserved_key(self, _):
+    keys = m.keys() - {key}
+    del m[key]
+    assert key not in m
+    assert len(m) < l1
+    # for k in keys:
+    #     assert k in m
```
Could we be more exact here and `assert len(m) == l1 - 1`?

Also, is there something strange breaking with the for loop so that it's currently commented out?

AnOctopus replied:
The for loop actually doesn't hold for session state as it currently functions, because for keyed widgets we have to represent it in `.keys()` as either the key or the widget id, but then deleting via the other one will cause it to no longer be present in the dict

vdonato commented on lib/tests/streamlit/state/session_state_test.py:
```diff
@@ -280,6 +283,7 @@ def test_date_input_serde(self):
         )
         check_roundtrip("date_interval", date_interval)
 
+    @pytest.mark.skip
```
Is this test in the process of being revived? Or is there something particularly difficult to work with in the new world that'll force us to revisit the file_uploader serde test later

vdonato commented on lib/tests/streamlit/state/strategies.py:
```diff
@@ -0,0 +1,102 @@
+    state.compact_state()
+    # round 2
+
+    new_state = draw(new_session_state)
+    for k, v in new_state.items():
+        state[k] = v
+
+    unkeyed_widgets = draw(
+        hst.dictionaries(keys=unkeyed_widget_ids, values=hst.integers())
+    )
+    for wid, v in unkeyed_widgets.items():
+        state.set_unkeyed_widget(mock_metadata(wid, v), wid)
+
+    widget_key_val_triple = draw(
+        hst.lists(hst.tuples(hst.uuids(), user_key, hst.integers()))
+    )
+    k_wids = {
+        key: (as_keyed_widget_id(wid, key), val)
+        for wid, key, val in widget_key_val_triple
+    }
+    for key, (wid, val) in k_wids.items():
+        state.set_keyed_widget(mock_metadata(wid, val), wid, key)
+
+    if k_wids:
+        session_state_widget_entries = draw(
+            hst.dictionaries(
+                keys=hst.sampled_from(list(k_wids.keys())),
+                values=hst.integers(),
+            )
+        )
+        for k, v in session_state_widget_entries.items():
+            state[k] = v
```
nit: could this be generalized into a helper function given that the code is essentially duplicated from above?

AnOctopus replied:
I originally had it inlined because a good way of doing that didn't occur to me, but I more recently realized that I could either have the function just do mutation of the same state instance twice, or even add a method for merging two session state instances, so I'll do one of those to clean it up

vdonato reviewed (commented):
Overall LGTM -- I have a few comments that are either minor or requests for increased test coverage in a few places.

Given that there's a decent amount going on here, I also want to take a second pass at this before hitting +1 on it, so I'll read through this another time tomorrow before hitting approve.

AnOctopus committed:
Fix compact method call

AnOctopus committed:
Don't skip legacy hashing keras test

I don't recall why I felt the need to add it, but the test doesn't seem
to break due to dependency issues anymore.

AnOctopus committed:
Remove incorrect plural in method name

AnOctopus committed:
Fix method names in wstate tests

AnOctopus committed:
Make dict deletion length property more precise

AnOctopus committed:
Remove invalid property for session state tests

Due to the requirements of `.keys()`, a keyed widget must be represented
by just one of {key, widget id}, and deleting a keyed widget using the
other one will cause the represented key to also be removed.

AnOctopus committed:
Update and format `_get_widget_id` docstring

AnOctopus committed:
Keep comments to roughly 80 char limit

AnOctopus committed:
Session state strategy refactoring

Using update methods, which must be custom because the default from
MutableMapping doesn't do the right thing for these types.

AnOctopus committed:
Fix test for `should_set_frontend_state_value`

AnOctopus committed:
Merge branch 'develop' into widgets-simplified-prototype

AnOctopus committed:
Mypy ignore WStates+SessionState update types

lawilby commented on lib/streamlit/state/session_state.py:
```diff
@@ -79,8 +73,8 @@ class Value:
     id: str
-    deserializer: WidgetDeserializer
-    serializer: WidgetSerializer
+    deserializer: WidgetDeserializer = attr.ib(repr=False)
```
@AnOctopus I am just learning about the attr library after reading your code. I am wondering why you are only using `attr.ib` here for `deserializer` and `serializer` and not for `id`, `value_type` and the rest.

AnOctopus replied:
the others happen automatically because `auto_attribs` is set to true in the decorator. but these two I use it explicitly so I can have them removed from the generated repr

lawilby commented on lib/streamlit/state/session_state.py:
```diff
@@ -429,63 +514,84 @@ def cull_nonexistent(self, widget_ids: Set[str]):
-            return deepcopy(metadata.deserializer(None, metadata.id))
+        """Get the value of a widget, for use as its return value.
+
+        This may be an existing value, or it may be generated if none exists.
```
I am wondering where the value is generated if none exists?

AnOctopus replied:
Ah, that should be updated, because this function no longer tries to do that

lawilby commented on lib/streamlit/state/widgets.py:
```diff
@@ -268,16 +271,14 @@ def _get_widget_id(
-    If user_key is defined, the widget_id returned is simply user_key.
-    Otherwise, we return a hash of the widget element type and the
-    string-serialized widget proto.
+    The widget id includes the user_key, which may be `None`, so widgets with
```
I don't fully understand this comment. If we have two of the same widgets with the same arguments and the user doesn't provide a user_key will the widget ids be the same?

AnOctopus replied:
Yes (and this will result in an exception when the second is created)

lawilby replied:
OK, that makes sense!

lawilby commented on lib/streamlit/state/widgets.py:
```diff
@@ -268,16 +271,14 @@ def _get_widget_id(
-    If user_key is defined, the widget_id returned is simply user_key.
-    Otherwise, we return a hash of the widget element type and the
-    string-serialized widget proto.
+    The widget id includes the user_key, which may be `None`, so widgets with
+    identical arguments can use it to be distinct.
+    The widget id includes an easily identified prefix, and the user_key as a
+    suffix, to make it easy to identify it and know if a key maps to it.
```
```suggestion
    The user_key is included in the widget id allowing widgets with identical arguments to have distinct ids. Note: user_key may be `None`. 
    The widget id includes an easily identified prefix, and the user_key as a
    suffix, to make it easy to identify it and know if a key maps to it.
```

Having the info about user_key possibly being `None` as a clause in the sentence made it less readable for me. Something like this might be a bit clearer.

AnOctopus committed:
Fix file uploader serde test

AnOctopus committed:
Update `get_value_for_registration` docstring

vdonato commented on lib/tests/streamlit/legacy_caching/hashing_test.py:
```diff
@@ -29,6 +29,7 @@
 from io import BytesIO
 from io import StringIO
 from unittest.mock import patch, Mock, MagicMock
+import pytest
```
nit: import is now unused

vdonato commented on lib/streamlit/state/session_state.py:
```diff
@@ -101,8 +95,9 @@ def __getitem__(self, k: str) -> Any:
-                    # No deserializer, which should only happen if state is gotten from a reconnecting browser
-                    # and the script is trying to access it. Pretend it doesn't exist.
+                    # No deserializer, which should only happen if state is
+                    # gotten from a reconnecting browser # and the script is
```
nit: there's an extra `#` in this line

vdonato commented on lib/streamlit/state/session_state.py:
```diff
@@ -260,23 +257,36 @@ class SessionState(MutableMapping[str, Any]):
-    _initial_widget_values: Dict[str, Any] = attr.Factory(dict)
+
+    # Keys used for widgets will be eagerly converted to the matching widget id
+    _key_id_mapping: Dict[str, str] = attr.Factory(dict)
```
nit: this might be slightly more clear if it were renamed `_key_wid_mapping` or `_key_to_wid_mapping` so that it's more obvious that we're going from a user-provided key to widget_id

vdonato commented on lib/streamlit/state/session_state.py:
```diff
@@ -297,45 +307,109 @@ def _safe_widget_state(self) -> Dict[str, Any]:
 
     @property
     def _merged_state(self) -> Dict[str, Any]:
```
Could this be removed entirely now? (along with `_safe_widget_state` above)

AnOctopus replied:
We use it for `__str__`

vdonato commented on lib/streamlit/state/session_state.py:
```diff
@@ -429,63 +514,87 @@ def cull_nonexistent(self, widget_ids: Set[str]):
             k: v
             for k, v in self._old_state.items()
-            if (k in widget_ids or not k.startswith(GENERATED_WIDGET_KEY_PREFIX))
+            if (k in widget_ids or not is_widget_id(k))
```
nit: the comment above is now slightly stale as we're removing all `widget_id`s in `_old_state` not present in the `widget_ids` set

vdonato commented on lib/streamlit/state/session_state.py:
```diff
@@ -297,45 +307,109 @@ def _safe_widget_state(self) -> Dict[str, Any]:
 
     def __getitem__(self, key: str) -> Any:
+        wid_key_map = self.reverse_key_wid_map
+        widget_id = self._get_widget_id(key)
+        if widget_id in wid_key_map and widget_id == key:
```
It took me a minute to understand what the significance of the `widget_id == key` clause is here, so it may be helpful to add a comment explaining that if this is the case, we were given a `widget_id` to begin with

vdonato reviewed (approved):
A few more small things that I noticed when taking a second pass at this (mostly nits / spotted what I think is some dead code that can be removed), but overall LGTM!

AnOctopus committed:
Test keyed widget duplicate ids

AnOctopus committed:
Remove unused import

AnOctopus committed:
Remove extra `#` in comment

AnOctopus committed:
Add comment for SessionState item lookup

AnOctopus committed:
Update `cull_nonexistent` comments

AnOctopus committed:
Clarify widget id generation docstring
